File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        return optimal(nums);
    }

    public int[] optimal(int[] nums){
        int[] ans = new int[nums.length];

        for(int i=0, product=1;i<ans.length;i++){
            if(i>0) product *= nums[i-1];
            ans[i] = product;
        }

        
        for(int i=ans.length-1, product=1;i>=0;i--){
            if(i < ans.length-1) product*= nums[i+1];
            ans[i] *= product;
        }   

        return ans;
    }


    public int[] bruteForce(int[] nums) {
        int[] ans = new int[nums.length];

        for(int i=0;i<nums.length;i++){
            ans[i] = 1;
            for(int j=0;j<nums.length;j++){
                if(i!=j){
                    ans[i] *= nums[j];
                }
            }
        }

        return ans;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        zeros = 0
        prod = 1
        for i in range(0,len(nums)):
            n = nums[i]
            if n==0:
                zeros+=1
            else:
                prod *= n

        res = [0]*len(nums)
        if zeros <= 1:
            for i in range(0,len(nums)):
                if zeros == 1:
                    if nums[i]==0:
                        res[i] = prod
                else:
                    res[i] = int(prod/nums[i])
        return res

        

        